### PR TITLE
Dance/fastapi

### PR DESCRIPTION
**Pull request by abjjabjj**
implementing CRUD functionality for clients, tools, and projects

Files Changed:
- `.pre-commit-config.yaml`
- `concrete/db/crud.py`
- `concrete/db/orm/__init__.py`
- `concrete/db/orm/models.py`
- `concrete/db/orm/setup.py`
- `pyproject.toml`
- `webapp/api/models.py`
- `webapp/api/server.py`

to run locally:
```make local-api```
and navigate to
```localhost:8000/docs```

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (abjjabjj) [inline .pre-commit-config.yaml:1]: Content Change

File: concrete/db/orm/__init__.py
Comment (abjjabjj) [inline concrete/db/orm/__init__.py:1]: Content Change

File: concrete/db/orm/models.py
Comment (abjjabjj) [inline concrete/db/orm/models.py:1]: Content Change

File: concrete/db/orm/setup.py
Comment (abjjabjj) [inline concrete/db/orm/setup.py:1]: Content Change

File: pyproject.toml
Comment (abjjabjj) [inline pyproject.toml:1]: Content Change

File: webapp/api/crud.py
Comment (abjjabjj) [inline webapp/api/crud.py:1]: Content Change

File: webapp/api/database.py
Comment (abjjabjj) [inline webapp/api/database.py:1]: Content Change

File: webapp/api/models.py
Comment (abjjabjj) [inline webapp/api/models.py:1]: Content Change

File: webapp/api/schemas.py
Comment (abjjabjj) [inline webapp/api/schemas.py:1]: Content Change

File: webapp/api/server.py
Comment (abjjabjj) [inline webapp/api/server.py:1]: Content Change

File: concrete/db/orm/schemas.py
Comment (abjjabjj) [inline concrete/db/orm/schemas.py:1]: Content Change

File: concrete/models/base.py [hunk 1]
```diff
@@ -8,7 +8,7 @@ class ConcreteModel(PydanticModel):
     def __str__(self):
```
Comment (MichaelDeng03) [inline concrete/models/base.py:8]: Is this going to be replaced orm Base eventually?

**Comment by kgang** [comment]
Whats going on with these comments? Also is this pr still draft/wip?

**Comment by abjjabjj** [comment]
Context for all the "Content Changes": when I was linting Vishnu's work I ended up linting all of concrete. Only the files listed in the initial comment and those with "Content Change" attached to them need to be looked at.

And this is still WIP, I haven't added messages yet

File: webapp/api/crud.py [hunk 2]
```diff
@@ -1,44 +1,125 @@
...
+
+# ===Operator=== #
+
+
+def create_operator(db: Session, operator_create: OperatorCreate) -> Operator:
+    operator = Operator(**operator_create.model_dump())
+    db.add(operator)
+    db.commit()
+    db.refresh(operator)
+    return operator
```
Comment (MichaelDeng03) [inline webapp/api/crud.py:37]: Return a sql response instead of operator to indicate that operator was correctly added?

File: concrete/models/base.py (hunk 1 above)
Comment (abjjabjj) [reply concrete/models/base.py:8]: It's uncertain to me currently. One the one hand, a merger might make the code more readable, but on the other hand not coupling everything to `SQLModel` makes the code less brittle.

**Comment by kgang** [comment]
Thanks for putting this up :nerd_face:

File: webapp/api/crud.py (hunk 2 above)
Comment (abjjabjj) [reply webapp/api/crud.py:37]: what kind of sql response? Are you thinking of some sort of wrapper? If the operator is successfully created then it has to be returned so the user knows what metadata is attached to it, most importantly its ID [](edited-with-Axolo)

**Comment by abjjabjj** [comment]
you got it boss

File: webapp/api/crud.py (hunk 2 above)
Comment (MichaelDeng03) [reply webapp/api/crud.py:37]: I'm not too clear on sqlalchemy. Does `return operator` reflect whatever db changes were made, like ID? And if it wasn't successfully created, then `return operator` wouldn't have ID/be None or something?

File: concrete/db/orm/models.py [hunk 3]
```diff
@@ -1,53 +1,81 @@
...
-        cascade="all, delete-orphan",
-    )
-    tools: Mapped[list["Tool"]] = relationship(
+
+class OperatorBase(Base):
+    instructions: str = Field(description="Instructions and role of the operator.")
+    title: str = Field(description="Title of the operator.", max_length=32)
+
+
+class OperatorUpdate(Base):
```
Comment (kgang) [inline concrete/db/orm/models.py:23]: Are you thinking that this is for changing operator instructions on the fly?

I might be inclined to spin up a new operator instead

File: concrete/db/orm/models.py
Comment (kgang) [reply concrete/db/orm/models.py:1]: Ooh file comments must be new 😯

File: webapp/api/crud.py (hunk 2 above)
Comment (abjjabjj) [reply webapp/api/crud.py:37]: `db.refresh(operator)` will update the model with the db changes. If an error occurs an exception will be thrown [](edited-with-Axolo)

File: concrete/db/orm/models.py (hunk 3 above)
Comment (kgang) [reply concrete/db/orm/models.py:23]: I think as far as changes operators need to support is changing association with tools or orchestrations but it might be sufficient to have that on those other models?

Half baked thoughts

File: pyproject.toml [hunk 4]
```diff
@@ -16,6 +16,7 @@ astor = "==0.8.1"
 celery = "==5.4.0"
 sqlalchemy = "==2.0.34"
 alembic = "==1.13.2"
+sqlmodel = "^0.0.22"
```
Comment (kgang) [inline pyproject.toml:19]: Lock

**Review by kgang — COMMENTED** [review]
Will review more soon

File: concrete/db/orm/models.py (hunk 3 above)
Comment (MichaelDeng03) [reply concrete/db/orm/models.py:23]: Maybe for self-improvement purposes it would be useful to be able to update an Operator?

Comment (kgang) [reply concrete/db/orm/models.py:23]: That is an interesting point. I'm inclined to think of an operator to be a model + instructions

File: concrete/models/clients.py [hunk 5]
```diff
@@ -8,12 +8,12 @@
 
 
 class OpenAIClientModel(ConcreteModel, KombuMixin):
-    model: str = Field(default='gpt-4o-mini', description='Name of LLM Model')
-    temperature: float = Field(default=0, description='Temperature of LLM Model')
+    model: str = Field(default="gpt-4o-mini", description="Name of LLM Model")
```
Comment (kgang) [inline concrete/models/clients.py:11]: nit: I feel like if we can set the default here than we shouldn't have it on the db sql model spec

File: concrete/db/orm/models.py (hunk 3 above)
Comment (abjjabjj) [reply concrete/db/orm/models.py:23]: There's still a need to change the title of the operator, so in that respect an update functionality is still necessary.

**Review by kgang — COMMENTED** [review]
Really nice stuff

If you could include info in the PR description next time when your ready to review with a vignette of what's possible (like a mini-tutorial) that'd help jumping in to run the code and sanity check everything

Not expecting this to get merged today at this point, but I can give another review tomorrow at the latest!

File: concrete/models/clients.py (hunk 5 above)
Comment (abjjabjj) [reply concrete/models/clients.py:11]: We can interact with the db sql model spec independently of this model. Are you suggesting we merge the two?

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -18,6 +18,9 @@ repos:
     rev: v1.11.2
     hooks:
       - id: mypy
+        additional_dependencies:
+        - "pydantic==v2.9.1"
+        - "sqlmodel==0.0.22"
```
[the remark this answers is not in the document]
Comment (abjjabjj) [reply .pre-commit-config.yaml:23]: Seems so: https://github.com/fastapi/sqlmodel/discussions/732

File: concrete/__main__.py [hunk 7]
```diff
@@ -11,14 +11,34 @@
...
+    "--celery",
+    action="store_true",
+    help="Use celery for processing. Otherwise, use plain.",
+)
 
 deploy_parser = subparsers.add_parser("deploy", help="Deploy image URIs to AWS")
-deploy_parser.add_argument("--image_uri", type=str, nargs='+', required=True, help="The image URIs to deploy to AWS")
 deploy_parser.add_argument(
-    "--container_name", type=str, nargs='+', required=True, help="The custom names for the containers"
+    "--image_uri",
```
[the remark this answers is not in the document]
Comment (abjjabjj) [reply concrete/__main__.py:22]: i can replace all of them with an update

**Comment by abjjabjj** [comment]
Will keep that in mind

But also I didn't intend for this to be reviewed yet since it's WIP; good comments nonetheless

**Comment by kgang** [comment]
You can leave us off as reviewers if you didn't want it, but I find reviews half way through speed up the final review

**Comment by kgang** [comment]
Also good chatting with the lads before the EOD :)

**Comment by abjjabjj** [comment]
gotcha gotcha

**Comment by abjjabjj** [comment]
good work lads

File: concrete/models/clients.py (hunk 5 above)
Comment (kgang) [reply concrete/models/clients.py:11]: Yeah I was saying the db model probably doesn't need a default but this one could stay [](edited-with-Axolo)

**Comment by kgang** [comment]
Good work mister lad

File: concrete/models/clients.py (hunk 5 above)
Comment (abjjabjj) [reply concrete/models/clients.py:11]: So you're also saying that if a user creates a client through the API they have to specify the fields, correct? [](edited-with-Axolo)

Comment (kgang) [reply concrete/models/clients.py:11]: Hm the api could have a default but the db probably shouldn't. I'm not totally certain what the best way to reason about it is [](edited-with-Axolo)

**Comment by kgang** [comment]
In the RESTful api at `:8000/docs` I created an Orchestrator than used its id to create an Operator, then used those IDs to create a client. Then I shut off the server and copy-pasted the sqllite server into my notebooks directory. I was able to look up my Operator using this code
```
from uuid import UUID
from concrete.db.orm.models import Operator
from sqlmodel import select

with SessionLocal() as session:
    stmt = select(Operator).where(Operator.id == UUID("b88bda71-4c7b-46bf-8087-87ceb22e363e"))
    print(session.scalars(stmt).first())
```
The crud's `get_operator` also fetched it

File: concrete/db/orm/models.py [hunk 8]
```diff
@@ -1,69 +1,252 @@
...
+from .setup import engine
 
-class Base(DeclarativeBase):
-    @declared_attr.directive
-    def __tablename__(cls) -> str:
-        return cls.__name__.lower()
-
-    id: Mapped[UUID] = mapped_column(UUID, primary_key=True, default=uuid4)
 
+class Base(SQLModel):
```
Comment (MichaelDeng03) [inline concrete/db/orm/models.py:10]: looks like we don't need to define the pydantic schemas anymore?

File: webapp/api/models.py
Comment (MichaelDeng03) [inline webapp/api/models.py:1]: is this file still necessary?

**Comment by MichaelDeng03** [comment]
```
cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) table operator has no column named orchestrator_id
[SQL: INSERT INTO operator (id, instructions, title, orchestrator_id) VALUES (?, ?, ?, ?)]
[parameters: ('e0f2e61df7ee42a7a588b306c2c05267', 'string', 'string', '22b644a4c72743288a11480678d223cf')]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
```

File: concrete/db/orm/models.py (hunk 8 above)
Comment (abjjabjj) [reply concrete/db/orm/models.py:10]: `SQLModel` essentially combines the pydantic schemas and the table definitions into a single class.

`schemas.py` no longer exists, but the models here are still necessary for FastAPI operation. Definitely open to suggestions on how to integrate the `db.orm.models` with those at root level.

**Comment by MichaelDeng03** [comment]
I tried making an orchestrator then an operator using that orchestrators id

File: webapp/api/models.py
Comment (abjjabjj) [reply webapp/api/models.py:1]: Yep, the singular model in it is used to make requests in the API

**Comment by MichaelDeng03** [comment]
:+1: works now

**Review by MichaelDeng03 — APPROVED** [review]
💯
